velox/exec/WindowPartition.cpp

```diff
@@ -407,26 +407,65 @@ void WindowPartition::updateKRangeFrameBounds(
             mappedFrameColumn,
             flags) == 0) {
       rawFrameBounds[i] = rawPeerBounds[i];
-    } else {
-      // If the search is for a preceding bound then rows between
-      // [0, currentRow] are examined. For following bounds, rows between
-      // [currentRow, numRows()) are checked.
-      if (isPreceding) {
+      continue;
+    }
+
+    auto compareWithLastRow = [&](vector_size_t currentRow) {
+      // For the first row, not to apply the sliding pattern.
+      if (currentRow == startRow) {
+        // For the first row, just returns a arbitrary positive value
+        // to indicate not to use sliding search pattern.
+        return 1;
+      }
+      return data_->compare(
+          partition_[currentRow - 1],
+          partition_[currentRow],
+          mappedFrameColumn,
+          mappedFrameColumn,
+          flags);
+    };
+
+    // If the search is for a preceding bound then rows between
+    // [0, currentRow] are examined. For following bounds, rows between
+    // [currentRow, numRows()) are checked. If the sliding search pattern
+    // is applicable, we also use the last row's search result to determine
+    // the start search bound for current row.
+
+    if (compareWithLastRow(currentRow) == 0) {
```

Comment:
We could have a local variable for compareWithLastRow(currentRow) as it is invoked multiple times in the code-path. Would be avoid to avoid data_->compare each time.